pageserver/src/walingest.rs

```diff
@@ -1419,13 +1419,36 @@ impl WalIngest {
                 forknum: VISIBILITYMAP_FORKNUM,
             };
 
-            let mut vm_page_no = rec.blkno / pg_constants::VM_HEAPBLOCKS_PER_PAGE;
-            if rec.blkno % pg_constants::VM_HEAPBLOCKS_PER_PAGE != 0 {
-                // Tail of last remaining vm page has to be zeroed.
-                // We are not precise here and instead of digging in VM bitmap format just clear the whole page.
-                modification.put_rel_page_image_zero(rel, vm_page_no)?;
+            //
+            // Code copied from PostgreSQL `visibilitymap_prepare_truncate` function in `visibilitymap.c`
+            //
+
+            // last remaining block, byte, and bit
+            let mut vm_page_no = rec.blkno / (pg_constants::VM_HEAPBLOCKS_PER_PAGE as u32);
+            let trunc_byte = rec.blkno as usize % pg_constants::VM_HEAPBLOCKS_PER_PAGE
+                / pg_constants::VM_HEAPBLOCKS_PER_BYTE;
+            let trunc_offs = rec.blkno as usize % pg_constants::VM_HEAPBLOCKS_PER_BYTE
+                * pg_constants::VM_BITS_PER_HEAPBLOCK;
+
+            // Unless the new size is exactly at a visibility map page boundary, the
+            // tail bits in the last remaining map page, representing truncated heap
+            // blocks, need to be cleared. This is not only tidy, but also necessary
+            // because we don't get a chance to clear the bits if the heap is extended
+            // again.
+            if (trunc_byte != 0 || trunc_offs != 0)
+                && self.shard.is_key_local(&rel_block_to_key(rel, vm_page_no))
+            {
+                modification.put_rel_wal_record(
+                    rel,
+                    vm_page_no,
+                    NeonWalRecord::TruncateVisibilityMap {
+                        trunc_byte,
+                        trunc_offs,
+                    },
+                )?;
```

Comment:
If the VM is already smaller, this can also create a TruncateVisibilityMap record against a page that doesn't exist. I think you'll get an error when it's replayed.

Please create a test case for that too.

Reply:
Sorry, I do not understand how it is possible that VM is already smaller.
We generate `TruncateVisibilityMap`for not completely truncated VM page. So at this moment it definitely exists. How can it disappear?

Reply:
Oh I see, Postgres would not have created the truncate record for the VM in the first place, if the page didn't exist. Got it.